File: src/app/persistence.rs
/// Persistence module for saving/loading project state.
///
/// Serializes the `AppState` to/from JSON files using the existing
/// serde derives already on `AppState`, `ConverterParams`, etc.

use crate::app::state::AppState;
use anyhow::{Context, Result};
use std::fs;

/// Default project file name
#[allow(dead_code)]
pub const DEFAULT_PROJECT_PATH: &str = "project.tupan.json";

/// Save the current project state to a JSON file.
pub fn save_project(path: &str, state: &AppState) -> Result<()> {
    let data = serde_json::to_string_pretty(state)
        .with_context(|| format!("failed to serialize project to {}", path))?;

    // Create backup of existing file if present
    if fs::metadata(path).is_ok() {
        let backup_path = format!("{}.bak", path);
        let _ = fs::copy(path, &backup_path);
    }

    fs::write(path, data)
        .with_context(|| format!("failed to write project to {}", path))?;

    Ok(())
}

/// Load a project state from a JSON file.
pub fn load_project(path: &str) -> Result<AppState> {
    let data = fs::read_to_string(path)
        .with_context(|| format!("failed to read project from {}", path))?;

    let mut state: AppState = serde_json::from_str(&data)
        .with_context(|| format!("failed to parse project from {}", path))?;

    // After loading, recalculate to ensure results are consistent
    state.recalculate();

    Ok(state)
}

/// Export the current schematic as SVG to a file.
pub fn export_schematic_svg(
    path: &str,
    elements: &[crate::schematic::primitives::SchematicElement],
) -> Result<()> {
    let svg = crate::schematic::export_svg::export_svg(elements, 500.0, 300.0);
    fs::write(path, svg)
        .with_context(|| format!("failed to write SVG to {}", path))?;
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::app::state::AppState;
    use std::fs;

    #[test]
    fn test_save_load_roundtrip() {
        let path = "/tmp/test_tupan_project.json";

        // Clean up
        let _ = fs::remove_file(path);

        let state = AppState::default();

        // Save
        save_project(path, &state).expect("save should succeed");

        // Load
        let loaded = load_project(path).expect("load should succeed");

        // Verify
        assert_eq!(state.active_converter, loaded.active_converter);
        assert_eq!(state.params.vin, loaded.params.vin);
        assert_eq!(state.params.vout_target, loaded.params.vout_target);
        assert_eq!(state.params.frequency, loaded.params.frequency);

        // Clean up
        let _ = fs::remove_file(path);
    }

    #[test]
    fn test_load_nonexistent_file() {
        let result = load_project("/tmp/nonexistent_tupan_file.json");
        assert!(result.is_err());
    }
}

File: tupan/app/persistence.py
"""Persistence module for saving/loading project state."""

import json
import os
from tupan.app.state import AppState
from tupan.domain import ConverterResults


def state_to_dict(state: AppState) -> dict:
    """Convert AppState to a JSON-serializable dict."""
    return {
        "status_message": state.status_message,
        "show_numerical_sim": state.show_numerical_sim,
        "design": {
            "vin": state.design.vin,
            "vout": state.design.vout,
            "duty_cycle": state.design.duty_cycle,
            "frequency": state.design.frequency,
            "iout_max": state.design.iout_max,
            "delta_il_pct": state.design.delta_il_pct,
            "delta_vo_pct": state.design.delta_vo_pct,
        },
        "results": {
            "vout": state.results.vout,
            "iout": state.results.iout,
            "iin": state.results.iin,
            "vout_ripple": state.results.vout_ripple,
            "il_ripple": state.results.il_ripple,
            "conduction_losses": state.results.conduction_losses,
            "switching_losses": state.results.switching_losses,
            "efficiency": state.results.efficiency,
        },
    }


def dict_to_state(d: dict) -> AppState:
    """Restore AppState from a JSON-deserialized dict."""
    state = AppState()

    state.status_message = d.get("status_message", "Loaded")
    state.show_numerical_sim = d.get("show_numerical_sim", False)

    des = d.get("design", {})
    state.design.vin = des.get("vin", 48.0)
    state.design.vout = des.get("vout", 12.0)
    state.design.duty_cycle = des.get("duty_cycle", 0.25)
    state.design.frequency = des.get("frequency", 100_000.0)
    state.design.iout_max = des.get("iout_max", 5.0)
    state.design.delta_il_pct = des.get("delta_il_pct", 0.30)
    state.design.delta_vo_pct = des.get("delta_vo_pct", 0.01)

    # Recompute from saved design params
    state.recalculate()

    # Optionally override results from saved values
    res = d.get("results", {})
    if res:
        state.results.vout = res.get("vout", state.results.vout)
        state.results.iout = res.get("iout", state.results.iout)
        state.results.iin = res.get("iin", state.results.iin)
        state.results.vout_ripple = res.get("vout_ripple", state.results.vout_ripple)
        state.results.il_ripple = res.get("il_ripple", state.results.il_ripple)
        state.results.conduction_losses = res.get("conduction_losses", state.results.conduction_losses)
        state.results.switching_losses = res.get("switching_losses", state.results.switching_losses)
        state.results.efficiency = res.get("efficiency", state.results.efficiency)

    return state


def save_project(path: str, state: AppState) -> None:
    """Save the current project state to a JSON file."""
    data = state_to_dict(state)
    if os.path.isfile(path):
        backup_path = path + ".bak"
        try:
            os.replace(path, backup_path)
        except OSError:
            pass
    with open(path, "w") as f:
        json.dump(data, f, indent=2)


def load_project(path: str) -> AppState:
    """Load a project state from a JSON file."""
    with open(path, "r") as f:
        data = json.load(f)
    return dict_to_state(data)
